platform/src/main/java/org/stellar/anchor/platform/action/NotifyOnchainFundsReceivedHandler.java

```diff
@@ -0,0 +1,136 @@
+package org.stellar.anchor.platform.action;
+
+import static org.stellar.anchor.api.platform.PlatformTransactionData.Kind.WITHDRAWAL;
+import static org.stellar.anchor.api.rpc.action.ActionMethod.NOTIFY_ONCHAIN_FUNDS_RECEIVED;
+import static org.stellar.anchor.api.sep.SepTransactionStatus.PENDING_ANCHOR;
+import static org.stellar.anchor.api.sep.SepTransactionStatus.PENDING_USR_TRANSFER_START;
+
+import java.util.Set;
+import javax.validation.Validator;
+import org.stellar.anchor.api.exception.AnchorException;
+import org.stellar.anchor.api.exception.rpc.InvalidParamsException;
+import org.stellar.anchor.api.exception.rpc.InvalidRequestException;
+import org.stellar.anchor.api.platform.PlatformTransactionData.Kind;
+import org.stellar.anchor.api.rpc.action.ActionMethod;
+import org.stellar.anchor.api.rpc.action.AmountRequest;
+import org.stellar.anchor.api.rpc.action.NotifyOnchainFundsReceivedRequest;
+import org.stellar.anchor.api.sep.SepTransactionStatus;
+import org.stellar.anchor.asset.AssetService;
+import org.stellar.anchor.horizon.Horizon;
+import org.stellar.anchor.platform.data.JdbcSep24Transaction;
+import org.stellar.anchor.platform.data.JdbcSepTransaction;
+import org.stellar.anchor.sep24.Sep24TransactionStore;
+import org.stellar.anchor.sep31.Sep31TransactionStore;
+
+public class NotifyOnchainFundsReceivedHandler
+    extends ActionHandler<NotifyOnchainFundsReceivedRequest> {
+
+  public NotifyOnchainFundsReceivedHandler(
+      Sep24TransactionStore txn24Store,
+      Sep31TransactionStore txn31Store,
+      Validator validator,
+      Horizon horizon,
+      AssetService assetService) {
+    super(
+        txn24Store,
+        txn31Store,
+        validator,
+        horizon,
+        assetService,
+        NotifyOnchainFundsReceivedRequest.class);
+  }
+
+  @Override
+  protected void validate(JdbcSepTransaction txn, NotifyOnchainFundsReceivedRequest request)
+      throws InvalidParamsException, InvalidRequestException {
+    super.validate(txn, request);
+
+    if (!((request.getAmountIn() == null
+            && request.getAmountOut() == null
+            && request.getAmountFee() == null)
+        || (request.getAmountIn() != null
+            && request.getAmountOut() != null
+            && request.getAmountFee() != null)
+        || (request.getAmountIn() != null
+            && request.getAmountOut() == null
+            && request.getAmountFee() == null))) {
```

Comment:
Can you clarify why this constraint was added?

Reply:
When Observer or Custody Server receives a notification about onchain payment, it should call Platform to update SEP transaction and provide `amountIn`(for example, when the user sent less or more than expected). In this case `amountOut` and `amountFee` are optional, because Anchor doesn't know exchange rates, however, they should also be updated, so the business will then call `notify_amounts_updated` action to update them. For offchain payments, there is no such logic because the business is responsible for tracking those payments and will provide all 3 amounts together if needed in `notify_offchain_funds_received` action